feat(hardware): added a script write and interact with the Flex eeprom.

vegano1 opened the pull request:
# Overview

We need to store the serial information for the Flex on the eeprom, this script will be used to write and verify that value along with any other data we might want to write to the eeprom.

Closes: [RET-1347](https://opentrons.atlassian.net/browse/RET-1347) & [RET-1349](https://opentrons.atlassian.net/browse/RET-1349)

# Test Plan

- [x] Write the serial number a Flex and verify that the value is stored after reboot

# Changelog

- added eeprom_writter.py script to read/write/print eeprom data on the Flex

# Review requests

- usage should be straightforward but let me know if there's anything we might need

# Risk assessment

Low

[RET-1347]: https://opentrons.atlassian.net/browse/RET-1347?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RET-1349]: https://opentrons.atlassian.net/browse/RET-1349?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

vegano1 committed:
eeprom module wip

vegano1 committed:
added ability to read property from eeprom

vegano1 committed:
switch read interface to Set instead of List
fix an issue when reading properties from the eeprom

vegano1 committed:
added raw serial number propId

vegano1 committed:
add property_write function and clean up

vegano1 committed:
Merge branch 'edge' into RET-1346_add_eeprom_module

vegano1 committed:
add gpio_dev to EEPROM object so we can set the eeprom write protect line SODIMM_222 low when writting to the eeprom

vegano1 committed:
remove swap files

vegano1 committed:
remove start/end tokens and implement TLV
deal with overflowing incomplete data
cleaning up

vegano1 committed:
- added filepath arg to EEPROM so we can override the default sys path for testing
- removed hard while loops when parsing data
- return set of PropertyId of the properties that were successfully written to the eeprom
- dropped the manufacturing_facility and unique_id values from the serial number
- removed unused PropertyId’s MACHINE_TYPE, MACHINE_VERSION, PROGRAMMED_DATE, UNIT_NUMBER
- added MAX_DATA_LEN so we can limit the size of a property to 255 bytes
- removed eeprom_writter.py script as this will be added in a separate pr
- added unit tests

vegano1 committed:
wip

vegano1 committed:
Merge branch 'edge' into RET-1346_add_eeprom_module

vegano1 committed:
created SystemDrivers holding class to make it easier for _build_system_hardware to return one item
added defines to drivers.__init__
use gpiod passed down from ot3Controller when initializing the EEPROMDriver

vegano1 committed:
change filepath to eeprom_path and make a Path type insted of string
return the EEPROMDriver module from the OT3Controller via a property
fix the eeprom_build_driver module

vegano1 committed:
Merge branch 'edge' into RET-1346_add_eeprom_module

vegano1 committed:
fix unit tests

vegano1 committed:
fixed an issue with data overflow
added unit tests

vegano1 committed:
Merge branch 'RET-1346_add_eeprom_module' into add-eeprom-writter-script

vegano1 committed:
_get_eeprom_info to get the name and size of the eeprom

vegano1 committed:
Merge branch 'RET-1346_add_eeprom_module' into RET-1349_eeprom_writter_script

vegano1 committed:
feat(hardware): added factory script to interact with the Flex eeprom.

sfoster1 commented on hardware/opentrons_hardware/scripts/eeprom_writter.py:
```diff
@@ -0,0 +1,293 @@
+"""This script is able to read/write to/from the Flex EEPROM.
```
nit: all of the detail about what arguments do should be in the argument help text instead of here so we don't have to write them twice

vegano1 replied:
We print out this module doc string using the __doc__ variable as part of the argParser description whenever we pass in --help.

sfoster1 replied:
Yeah but `--help` will also print out the helpstring for each argument on its own, so itll be duplicated

sfoster1 commented on hardware/opentrons_hardware/scripts/eeprom_writter.py:
```diff
@@ -0,0 +1,293 @@
+            "Need at least one valid property and value to write.\n"
+            "Pass in the '--help' arg for an example."
+        )
+        exit(1)
```
nit: in a script i generally prefer to get most things being exceptions rather than `sys.exit` so the functions are usable elsewhere

sfoster1 reviewed (changes requested):
`eeprom_writer` not `eeprom_writter` and some nits, ut otherwise looks good!

vegano1 committed:
fix naming typo
raise errors instead of calling exit()
add serial number format description
only write FORMAT_VERSION if we have at least one other property to write

laviera commented on hardware/opentrons_hardware/scripts/eeprom_writer.py:
```diff
@@ -0,0 +1,303 @@
+    # write
+    try:
+        for idx in range(pages):
+            data = b"\xff" * DEFAULT_READ_SIZE
```
this "data= " statement should be outside of the for loop

laviera commented on hardware/opentrons_hardware/scripts/eeprom_writer.py:
```diff
@@ -0,0 +1,303 @@
+    # clear the eeprom first
+    success, msg = clear_eeprom(eeprom_api)
+    if not success:
+        msg = f"Error writting to eeprom: Could not clear eeprom - {msg}"
```
nit pick - spelling writting == writing

vegano1 committed:
fix typos

vegano1 committed:
Merge branch 'edge' into RET-1349_eeprom_writter_script

laviera commented:
@vegano1 - Should there be a "print raw" - allowing us to just dump the raw eeprom

sfoster1 reviewed (approved):
Looks good overall!
